test/datastore-test.js

```diff
@@ -0,0 +1,851 @@
+'use strict';
+
+/* global describe, it */
+
+var chai = require('chai');
+var expect = chai.expect;
+var chaiAsPromised = require('chai-as-promised');
+var dataStore = require('../src/models/datastore');
+var models = require('./data/models/Models');
+var STORE_NAME = 'test';
+
+chai.use(chaiAsPromised);
+
+describe('models.DataStore', function () {
+  // Basic setup.
+  describe('DataStore.configure', function () {
+    it ('throws if no config is given', function () {
+      return expect(dataStore.configure).to.throw(
+        Error, 'DataStore configuration not provided.'
+      );
+    });
+  });
+
+  // Put.
+  describe('DataStore.put', function () {
+    // Set up the data store before each run.
+    beforeEach (function () {
+      dataStore.configure({
+        store: STORE_NAME,
+        models: models
+      });
+    });
+
+    // Delete the test store contents directly each time.
+    afterEach(function (done) {
+      let mongoose = require('mongoose');
+      mongoose.connect('mongodb://localhost/' + STORE_NAME, function () {
+        mongoose.connection.db.dropDatabase();
+        mongoose.disconnect();
+        done();
+      });
+    });
+
+    it ('rejects if no type is given', function () {
+      var put = dataStore.put(undefined, {});
+
+      return expect(put).to.eventually.be.rejectedWith(
+        Error, 'Model type not provided.'
+      );
+    });
+
+    it ('rejects if no data is given', function () {
+      var put = dataStore.put('test');
+
+      return expect(put).to.eventually.be.rejectedWith(
+        Error, 'put() requires an object.'
+      );
+    });
+
+    it ('rejects if data is not an object (requires)', function () {
+      var put = dataStore.put('testnorequire', 1);
+
+      return expect(put).to.eventually.be.rejectedWith(
+        Error, 'put() requires an object.'
+      );
+    });
+
+    it ('rejects if data is an array with no objects', function () {
+      var put = dataStore.put('testnorequire', [1, {
+        name: 'testObject'
+      }]);
+
+      return expect(put).to.eventually.be.rejectedWith(
+        Error, 'put() given array of entries which are not all objects.'
+      );
+    });
+
+    it ('rejects if data contains an empty property', function () {
```

Comment:
Sorry I can't follow this test.

For example:
var ConnectedModel = models.test.factory(connection);

models.test <- what is test, a model type?
models.test.factory <- I expect it to make a mode of somesort - but ConnectedModel I'm not sure what that is exactly.

var instance = new ConnectedModel(testObject);
instance.save(err, result);

You care instantiating a model of some sort, passing in a object and saving the instance. Saving to mongoosedb I presume, but what you are saving isn't clear - the testObject? Also where does the instance on the model occur?